Improve the Instructions iterator for scripts

stevenroose opened the pull request:
- Rename the `iter` method to `instructions`.
- Add `instructions_minimal` for minimal-enforced iteration.
- Iterator has `Result<Instruction, Error>` as items.

Fixes https://github.com/rust-bitcoin/rust-bitcoin/issues/396.

thomaseizinger commented on src/blockdata/script.rs:
```diff
@@ -329,10 +329,21 @@ impl Script {
+
+    /// Iterate over the script in the form of `Instruction`s while enforcing
+    /// minimal pushes.
+    pub fn instructions_minimal(&self) -> Instructions {
```
Nit: What do you think about naming this `minimal_instructions` instead? The motivation would be to be closer to how you would use it in an English sentence :)

dpc replied:
The advantage of current version is autocomplete-friendiness.

Alternative would be to have `fn enforce_minimal(self) - >Self` on `Instructions` itself, so it would be `instructions().enforce_minimal()`. Less discoverable, maybe, but reads better, and generalizes better in case further options would be introduced in the future.

stevenroose replied:
@thomaseizinger Hmm, I think the `<method>_<specialization thereof>()` pattern is common, no? Like `iter_mut` instead of `mut_iter` (as per mutable iterator)..

thomaseizinger replied:
Fair points :)

dr-orlovsky commented on src/blockdata/script.rs:
```diff
@@ -448,9 +457,9 @@ pub struct Instructions<'a> {
+    type Item = Result<Instruction<'a>, Error>;
 
-    fn next(&mut self) -> Option<Instruction<'a>> {
+    fn next(&mut self) -> Option<Result<Instruction<'a>, Error>> {
```
`Option<Result<...>>` is a strange construction as for `Iterator::next()`

For sure, there is a set of `try_*` methods for iterators, returning `Result`, however it's not the `iter()` itself which is supposed to be not only optional, but failable at the second level.

I would probably suggest to move from lazy script parsing to a pre-parsing, returning `Result`, and iterating over it. I can't imagine situation where the script is not consumed as whole, so lazy parsing does not make a lot of sense.

Thus, something like `script.instructions()?.iter().map(|command| ...)` will probably be more idiomatic.

dpc replied:
Pre-parsing would mean traversing the script twice. I don't know if the speed difference would be noticeable.

At least IO stuff in stdlib does `Item = Result<_>` - https://doc.rust-lang.org/std/io/trait.Read.html#method.bytes , though I must admit handling of such types is under-featured, and I often resolve to using https://docs.rs/resiter/0.4.0/resiter/ .

From PoV of the user I don't find having to handle `Instruction::Error` any better than having to handle `Result` in the `map`. At least with `Item = Result<_>` I can use `resiter`, or write my own reusable adapter, while for `Instruction::Error` it is not possible.

stevenroose replied:
Yeah you can either `?` in a loop, or collect if you want the vector
```rust
let instructions: Result<Vec<Instruction<'a>>, Error> = script.instructions().collect();
```
I'm not sure if the following would work:
```rust
let instructions = script.instructions().collect::<Result<Vec<Instruction<'a>>, Error>>().unwrap();
```

apoelstra commented:
travis is actually failing on 1.22 it looks like

dpc commented on src/blockdata/script.rs:
```diff
@@ -1008,55 +1017,53 @@ mod test {
-                Instruction::Error(Error::NonMinimalPush),
-            ]
+            v_nonmin.err().unwrap(),
+            Error::NonMinimalPush,
```
This trailing comma confuses the old rust compiler.

stevenroose commented:
Updated and rebased.

apoelstra commented:
Can you kick travis? Seems I cannot. Maybe it is just too old.

stevenroose commented:
Rebased and added this small nit:
```diff
diff --git a/src/util/contracthash.rs b/src/util/contracthash.rs
index 0d94b99..345e9bc 100644
--- a/src/util/contracthash.rs
+++ b/src/util/contracthash.rs
@@ -231,7 +231,7 @@ pub fn untemplate(script: &script::Script) -> Result<(Template, Vec<PublicKey>),
         if let Err(e) = instruction {
             return Err(Error::Script(e));
         }
-        match instruction.ok().unwrap() {
+        match instruction.unwrap() {
             script::Instruction::PushBytes(data) => {
                 let n = data.len();
                 ret = match PublicKey::from_slice(data) {
```
This should also kick Travis. I don't know what's going on but on ElementsProject/elements, I also can't restart Travis jobs anymore. It's been like that for a few weeks now.

thomaseizinger commented on src/util/contracthash.rs:
```diff
@@ -227,8 +227,11 @@ pub fn untemplate(script: &script::Script) -> Result<(Template, Vec<PublicKey>),
-    for instruction in script.iter(false) {
-        match instruction {
+    for instruction in script.instructions() {
+        if let Err(e) = instruction {
+            return Err(Error::Script(e));
+        }
+        match instruction.unwrap() {
```
Any reason to write this with unwrap and not use `?`?

```suggestion
        let instruction = instruction.map_err(Error::Script)?;
        
        match instruction {
```

If the appropriate `From` impl is in place, you can even omit the `map_err`.

stevenroose replied:
No particular reason. I might have had a reason for not having the `From` in place. And I think like this it might be more readable.

elichai replied:
I think `?` is nicer, but can be done in a future PR

apoelstra reviewed (approved):
This looks much better!

elichai reviewed (approved):
Looks good! it's nicer to use the `Result`
